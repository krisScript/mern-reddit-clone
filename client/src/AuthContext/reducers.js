export const LOGIN = 'LOGIN';
export const SIGNUP = 'SIGNIN';
export const LOGOUT = 'LOGOUT';

const login = async authData => {
  try {
    const { email, password } = authData;
    const response = await fetch('http://localhost:8080/auth/login', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        email,
        password
      })
    });
    if (response.status === 422) {
      throw new Error('Validation failed.');
    }
    if (response.status !== 200 && response.status !== 201) {
      console.log('Error!');
      throw new Error('Could not authenticate you!');
    }
    if (response.status === 200) {
      const responseData = await response.json();

      localStorage.setItem('token', responseData.token);
      localStorage.setItem('userId', responseData.userId);
      const remainingMilliseconds = 60 * 60 * 1000;
      const expiryDate = new Date(new Date().getTime() + remainingMilliseconds);
      localStorage.setItem('expiryDate', expiryDate.toISOString());
      // this.setAutoLogout(remainingMilliseconds);
      return {
        isAuth: true,
        token: responseData.token,
        userId: responseData.userId
      };
    }
  } catch (err) {
    console.log(err);
  }
};

const signup = async authData => {
  try {
    const { username, email, password, matchPassword } = authData;
    const response = await fetch('http://localhost:8080/auth/signup', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        username,
        email,
        password,
        matchPassword
      })
    });
    if (response.status === 422) {
      throw new Error(
        "Validation failed. Make sure the email address isn't used yet!"
      );
    }
    if (response.status !== 200 && response.status !== 201) {
      console.log('Error!');
      throw new Error('Creating a user failed!');
    } else if (response.status === 200) {
      const responseData = await response.json();
      this.props.history.replace('/');

      return {
        isAuth: true,
        token: responseData.token,
        userId: responseData.userId
      };
    }
  } catch (err) {
    console.log(err);
  }
};
const logout = () => {
  localStorage.removeItem('token');
  localStorage.removeItem('expiryDate');
  localStorage.removeItem('userId');
  return { isAuth: false, token: null, userId: null };
};
export const authReducer = (authData, action) => {
  switch (action.type) {
    case LOGIN:
      return login(authData);
    case SIGNUP:
      return signup(authData);
    case LOGOUT:
      return logout();
    default:
      return authData;
  }
};
