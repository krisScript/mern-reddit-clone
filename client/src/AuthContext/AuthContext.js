import React, { useState, useReducer, createContext,useEffect } from 'react';
import { BrowserRouter, Route, Link, Switch } from 'react-router-dom';
import { authReducer, LOGIN,LOGOUT} from './reducers';
import autoLogout from '../authUtil/autoLogout'
const Context = createContext();
const AuthContext = props => {
  const [authState, dispatch] = useReducer(authReducer, {
    isAuth: false,
    token: null,
    userId: null,
  });
  const loginReducer = (authData) => dispatch({ type:LOGIN,authData})
  const logoutReducer = () => dispatch({type:LOGOUT,authData:{}})
  useEffect(() => {
    const token = localStorage.getItem('token');
    const expiryDate = localStorage.getItem('expiryDate');
    if (!token || !expiryDate) {
      return;
    }
    if (new Date(expiryDate) <= new Date()) {
      logoutReducer()
      return;
    }
    const userId = localStorage.getItem('userId');
    const remainingMilliseconds =
      new Date(expiryDate).getTime() - new Date().getTime();
    loginReducer({isAuth:true,token,userId})
   
    // autoLogout(remainingMilliseconds,logoutReducer)
  },[])
  return (
    <Context.Provider
      value={{
        authState,
        loginReducer,
        logoutReducer
      }}
    >
      {props.children}
    </Context.Provider>
  );
};
export const AuthContextData = Context ;
export const AuthContextConsumer = Context.Consumer
export default AuthContext;
