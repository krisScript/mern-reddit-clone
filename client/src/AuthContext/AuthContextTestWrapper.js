import {AuthContextData} from './AuthContext'
import {BrowserRouter} from 'react-router-dom'
import React from 'react'
const AuthContextTestWrapper = ({authState,children,updateUserDataReducer}) => {
    return (
       <AuthContextData.Provider value={{authState,updateUserDataReducer}}>
        <BrowserRouter>
          {children}
          </BrowserRouter>
        </AuthContextData.Provider>
    )
}
export default AuthContextTestWrapper 