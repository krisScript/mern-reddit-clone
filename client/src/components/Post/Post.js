import React from 'react';
import { library } from '@fortawesome/fontawesome-svg-core';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { AuthContextData } from '../../AuthContext/AuthContext';
import {
  faShare
} from '@fortawesome/free-solid-svg-icons';
library.add(faShare);
const Post = props => {
  const { post } = props;
  const {title,author} = post
  return (
    <div  className="post">
    <h4 className="post-author">Posted by u/{author}</h4>
    <h3 className="post-title">{title}</h3>
    <div className="post-info">
    <button className="button post-info-button">
    <FontAwesomeIcon icon="share" />Share
    </button>
    <button className="button post-info-button">
    <FontAwesomeIcon icon="share" />Share
    </button>
    <button className="button post-info-button">
    <FontAwesomeIcon icon="share" />Share
    </button>
    <button className="button post-info-button">
    <FontAwesomeIcon icon="share" />Share
    </button>
    </div>
    </div>
  );
};
export default Post;
