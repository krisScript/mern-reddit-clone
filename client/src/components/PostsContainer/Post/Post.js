import React,{useContext} from 'react';
import {Link} from 'react-router-dom'
import { library } from '@fortawesome/fontawesome-svg-core';
import { AuthContextData } from '../../AuthContext/AuthContext';
import PostBar from '../PostBar/PostBar'
import VotesContainer from '../VotesContainer/VotesContainer'
const Post = props => {
  const authState = useContext(AuthContextData)
  const {isAuth,userId,token}  = authState.authState
  const { post,community,deletePostElement } = props;
  const {title,author,comments,communityName} = post
  console.log(post)
  return (
    <div  className="post">
    <VotesContainer post={post} />
    <div className="post-body">
    <div className="post-links">
    <Link to={`/community/${communityName}`} className="post-community">c/{communityName}</Link>
    <Link to={`/community/${communityName}`} className="post-author">c/{author}</Link>
    <Link  to={{pathname:`/post/${post._id}`, community}} className="post-content">
    <h3 className="post-title">{title}</h3>
    </Link>
    </div>
    <PostBar post={post} token={token} userId={userId} deletePostElement={deletePostElement} isAuth={isAuth} />
    </div>
    </div>
  );
};
export default Post;
