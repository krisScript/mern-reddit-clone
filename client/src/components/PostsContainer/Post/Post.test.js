import React from 'react';
import {BrowserRouter} from 'react-router-dom'
import { render, fireEvent, waitForElement } from 'react-testing-library';
import Post from './Post';


import {AuthContextData} from '../../../AuthContext/AuthContext'
const updateUserDataReducer = jest.fn()
const authState =  {isAuth:false,token:null,userId:null,userData:null,updateUserDataReducer}
describe('<Post/>',() => {
    const post = {
        title:'post',
        author:'NewUser',
        comments:0,
        communityName:'react',
        comments:0
    }
    const community = {
        communityName:'react'
    }
    const deletePostElement  = jest.fn()

    
    const { container, rerender,queryByPlaceholderText,queryByValue ,getByPlaceholderText,getByText } = render (<Post post={post} community={community} deletePostElement={deletePostElement} />,{wrapper: ({ children }) => (
        <AuthContextData.Provider value={{authState}}>
        <BrowserRouter>
          {children}
          </BrowserRouter>
        </AuthContextData.Provider>
      )})

    it('should have element with c/post.communityName text', async () => {
        const element =  getByText(`c/${post.communityName}`)
        expect(element.className).toMatch('post-community')
        expect(element).toBeTruthy;
      });
      it('should have element with u/post.author text', async () => {
        const element =  getByText(`u/${post.author}`)
        expect(element.className).toMatch('post-author')
        expect(element).toBeTruthy;
      });
      it('should have element with post.title text', async () => {
        const element =  getByText(`${post.title}`)
        expect(element.className).toMatch('post-title')
        expect(element).toBeTruthy;
      });
      
})

