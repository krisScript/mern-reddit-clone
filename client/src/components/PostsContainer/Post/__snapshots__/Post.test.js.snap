// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Post/> snapshot 1`] = `
<div>
  <div
    class="post"
  >
    <div
      class="votes-container"
    >
      <button
        class="button "
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-caret-up fa-w-10 "
          data-icon="caret-up"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 320 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M288.662 352H31.338c-17.818 0-26.741-21.543-14.142-34.142l128.662-128.662c7.81-7.81 20.474-7.81 28.284 0l128.662 128.662c12.6 12.599 3.676 34.142-14.142 34.142z"
            fill="currentColor"
          />
        </svg>
      </button>
      <span
        class="votes-container-number  "
      >
        NaN
      </span>
      <button
        class="button "
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-caret-down fa-w-10 "
          data-icon="caret-down"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 320 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
    <div
      class="post-body"
    >
      <div
        class="post-links"
      >
        <a
          class="post-community"
          href="/community/react"
        >
          c/
          react
        </a>
        <a
          class="post-author"
          href="/community/react"
        >
          u/
          NewUser
        </a>
        <a
          class="post-content"
          href="/post/undefined"
        >
          <h3
            class="post-title"
          >
            post
          </h3>
        </a>
      </div>
      <div
        class="post-info"
      >
        <button
          class="button post-info-button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-share fa-w-16 "
            data-icon="share"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M503.691 189.836L327.687 37.851C312.281 24.546 288 35.347 288 56.015v80.053C127.371 137.907 0 170.1 0 322.326c0 61.441 39.581 122.309 83.333 154.132 13.653 9.931 33.111-2.533 28.077-18.631C66.066 312.814 132.917 274.316 288 272.085V360c0 20.7 24.3 31.453 39.687 18.164l176.004-152c11.071-9.562 11.086-26.753 0-36.328z"
              fill="currentColor"
            />
          </svg>
          <span>
            Share
          </span>
        </button>
        <button
          class="button post-info-button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-comment-alt fa-w-16 "
            data-icon="comment-alt"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M448 0H64C28.7 0 0 28.7 0 64v288c0 35.3 28.7 64 64 64h96v84c0 9.8 11.2 15.5 19.1 9.7L304 416h144c35.3 0 64-28.7 64-64V64c0-35.3-28.7-64-64-64z"
              fill="currentColor"
            />
          </svg>
          <span>
            0
          </span>
        </button>
        <button
          class="button post-info-button  "
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-bookmark fa-w-12 "
            data-icon="bookmark"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 384 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 512V48C0 21.49 21.49 0 48 0h288c26.51 0 48 21.49 48 48v464L192 400 0 512z"
              fill="currentColor"
            />
          </svg>
          <span>
            Save
          </span>
        </button>
        <button
          class="button post-info-button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-copy fa-w-14 "
            data-icon="copy"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M320 448v40c0 13.255-10.745 24-24 24H24c-13.255 0-24-10.745-24-24V120c0-13.255 10.745-24 24-24h72v296c0 30.879 25.121 56 56 56h168zm0-344V0H152c-13.255 0-24 10.745-24 24v368c0 13.255 10.745 24 24 24h272c13.255 0 24-10.745 24-24V128H344c-13.2 0-24-10.8-24-24zm120.971-31.029L375.029 7.029A24 24 0 0 0 358.059 0H352v96h96v-6.059a24 24 0 0 0-7.029-16.97z"
              fill="currentColor"
            />
          </svg>
          <span>
            Spam
          </span>
        </button>
        
      </div>
    </div>
  </div>
</div>
`;
